rpcs3/Emu/Cell/SPUThread.cpp

```diff
@@ -1379,26 +1379,9 @@ void spu_thread::do_dma_transfer(const spu_mfc_cmd& args)
 				break;
 			}
 
+			assert(size & 0xf == 0);
 			auto lock = vm::passive_lock(eal & -128, ::align(eal + size, 128));
-
-			while (size >= 128)
-			{
-				mov_rdata(*reinterpret_cast<decltype(spu_thread::rdata)*>(dst), *reinterpret_cast<const decltype(spu_thread::rdata)*>(src));
-
-				dst += 128;
-				src += 128;
-				size -= 128;
-			}
-
-			while (size)
-			{
-				*reinterpret_cast<v128*>(dst) = *reinterpret_cast<const v128*>(src);
-
-				dst += 16;
-				src += 16;
-				size -= 16;
-			}
-
+			memmove(dst, src, size);
```

Comment:
There are three reasons as to why we had kept the implemrntation specialized:
- We control the transfers:
1. We control the code generated which the transfers are carried om almost exclusively, sometimes better than the compiler. A good example is as why we are usig mov_rdata im atomic mfc commanfs - previously MSVC generated code which takes bad use of sse registers access reordering
1. We are aware of hints memmove/memcpy won't know at first such as the fact that dize, src and dst are always 16 bytes long aligned, potentially leading to better optimizations. Even if they check so internally it is still processing time wasted on checks which shouldnf have hapoened.
- MFC transfers which which bigger or equal in size to 16 bytes, must use atleast atomic disjoonted accesses to read/write to main memory. std::memcpy and std::memmove may not allow this (although even currrent implementation isn't perfect regarding it, we may add an accuracy setting for this in the future).

Reply:
1. That sounds like a very dangerous assumption to make, compilers evolve, as do processors and libcs, and you will never have as much resource as them to catch up and test on a wide variety of systems.
2. Everything is in the “potentially”, in this case you clearly don’t (on my system). I tested the suggested AVX path, but it still goes for 3.28% total (compared to 3.74% currently, and 2.95% with this PR). `memmove()` on my Kaby Lake system uses `rep movsb` which has the most throughput when ran on a 32-aligned buffer according to agner fog’s guide, which is the case half of the time here. Don’t worry about trailer and header code, profile first and use the best compromise between code amount, portability and performances.
3. Hmm, do you have any game which relies on this race condition, or is it purely theoretical?